core/state/trie_prefetcher.go

```diff
@@ -276,365 +216,175 @@ func (p *triePrefetcher) trieID(owner common.Hash, root common.Hash) string {
 	}
-	sf.to.wait()
 }
 
+// abort interrupts the subfetcher immediately. It is safe to call abort multiple
+// times but it is not thread safe.
 func (sf *subfetcher) abort() {
-	if sf.to == nil {
-		// Unable to open trie
-		return
-	}
-	sf.to.abort()
-}
-
-func (sf *subfetcher) skips() int {
-	if sf.to == nil {
-		// Unable to open trie
-		return 0
-	}
-	return sf.to.skipCount()
-}
-
-func (sf *subfetcher) copies() int {
-	if sf.to == nil {
-		// Unable to open trie
-		return 0
+	select {
+	case <-sf.stop:
+	default:
+		close(sf.stop)
 	}
-	return sf.to.copies
+	<-sf.term
 }
 
-// trieOrchestrator is not thread-safe.
-type trieOrchestrator struct {
-	sf *subfetcher
-
-	// base is an unmodified Trie we keep for
-	// creating copies for each worker goroutine.
-	//
-	// We care more about quick copies than good copies
-	// because most (if not all) of the nodes that will be populated
-	// in the copy will come from the underlying triedb cache. Ones
-	// that don't come from this cache probably had to be fetched
-	// from disk anyways.
-	base     Trie
-	baseLock sync.Mutex
-
-	tasksAllowed bool
-	skips        int // number of tasks skipped
-	pendingTasks [][]byte
-	taskLock     sync.Mutex
-
-	processingTasks sync.WaitGroup
-
-	wake     chan struct{}
-	stop     chan struct{}
-	stopOnce sync.Once
-	loopTerm chan struct{}
-
-	copies      int
-	copyChan    chan Trie
-	copySpawner chan struct{}
-}
+// loop waits for new tasks to be scheduled and keeps loading them until it runs
+// out of tasks or its underlying trie is retrieved for committing.
+func (sf *subfetcher) loop() {
+	// No matter how the loop stops, signal anyone waiting that it's terminated
+	defer func() {
+		sf.pool.wait()
+		close(sf.term)
+	}()
 
-func newTrieOrchestrator(sf *subfetcher) *trieOrchestrator {
 	// Start by opening the trie and stop processing if it fails
-	var (
-		base Trie
-		err  error
-	)
 	if sf.owner == (common.Hash{}) {
-		base, err = sf.db.OpenTrie(sf.root)
+		trie, err := sf.db.OpenTrie(sf.root)
 		if err != nil {
 			log.Warn("Trie prefetcher failed opening trie", "root", sf.root, "err", err)
-			return nil
+			return
 		}
+		sf.trie = trie
 	} else {
 		// The trie argument can be nil as verkle doesn't support prefetching
 		// yet. TODO FIX IT(rjl493456442), otherwise code will panic here.
-		base, err = sf.db.OpenStorageTrie(sf.state, sf.addr, sf.root, nil)
+		trie, err := sf.db.OpenStorageTrie(sf.state, sf.addr, sf.root, nil)
 		if err != nil {
 			log.Warn("Trie prefetcher failed opening trie", "root", sf.root, "err", err)
-			return nil
+			return
 		}
+		sf.trie = trie
 	}
+	// Trie opened successfully, keep prefetching items
+	for {
+		select {
+		case <-sf.wake:
+			// Subfetcher was woken up, retrieve any tasks to avoid spinning the lock
+			sf.lock.Lock()
+			tasks := sf.tasks
+			sf.tasks = nil
+			sf.lock.Unlock()
+
+			// Prefetch any tasks until the loop is interrupted
+			for _, task := range tasks {
+				select {
```

Comment:
Don't we want to handle the `case <-sf.stop:` as well?

Reply:
we don't, we modified it to wait for all requests to process.